minetest/minetest
# Catch C++ -> Lua exceptions with PUC Lua

Add compact, short information about your PR for easier understanding:

- Goal of the PR
  - This PR aims to bring Lua API exception handling with PUC Lua in line with LuaJIT.
- How does the PR work?
  - A wrapper macro is added. It takes a `lua_CFunction` and returns a `lua_CFunction` that handles C++ exceptions. With LuaJIT this macro is a no-op. It is called on all Lua API functions.
  - Compiling Lua is C++ solves a separate issue from this. Without this PR, exceptions are caught, but the error object is a function instead of what it should be: the error message.
- Does it resolve any reported issue?
  - No.
- Does this relate to a goal in [the roadmap](../doc/direction.md)?
  - No.
- If not a bug fix, why is this PR needed? What usecases does it solve?
  - This is a bug fix.

## To do

This PR is Ready for Review.

## How to test

1. Compile Minetest using the bundled Lua.
2. Enable devtest unit testing.
3. Put this code into a mod:
    ```lua
    local succeeded, message = pcall(minetest.get_content_id, "jkfdkjfkj")
    if not succeeded then minetest.log("error", message) end
    ```
3. Load a devtest world with the mod enabled.
4. The unit tests should pass.
5. You should see the error message generated by `minetest.get_content_id` but the world should not crash on startup.
